File: sdks/go/crypto.go
package agentconfig

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"regexp"
)

const (
	SaltBytes   = 16
	IVBytes     = 12
	KeyBytes    = 32
	GCMTagBytes = 16
)

var (
	base64Pattern    = regexp.MustCompile(`^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$`)
	base64URLPattern = regexp.MustCompile(`^[A-Za-z0-9_-]*$`)
)

func validatePassword(password string) error {
	if len(password) == 0 {
		return errors.New("password must not be empty")
	}
	return nil
}

func validateIterations(iterations int) error {
	if iterations < PBKDF2MinIterations || iterations > PBKDF2MaxIterations {
		return fmt.Errorf("PBKDF2 iterations must be between %d and %d", PBKDF2MinIterations, PBKDF2MaxIterations)
	}
	return nil
}

func toB64(data []byte) string {
	return base64.StdEncoding.EncodeToString(data)
}

func fromB64(value string, fieldName string) ([]byte, error) {
	if !base64Pattern.MatchString(value) {
		return nil, fmt.Errorf("invalid %s", fieldName)
	}
	decoded, err := base64.StdEncoding.DecodeString(value)
	if err != nil {
		return nil, fmt.Errorf("invalid %s: %w", fieldName, err)
	}
	if toB64(decoded) != value {
		return nil, fmt.Errorf("non-canonical %s", fieldName)
	}
	return decoded, nil
}

func toB64URL(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}

func fromB64URL(value string) ([]byte, error) {
	if !base64URLPattern.MatchString(value) || len(value)%4 == 1 {
		return nil, errors.New("invalid bundle base64url payload")
	}
	decoded, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		return nil, fmt.Errorf("invalid base64url payload: %w", err)
	}
	if toB64URL(decoded) != value {
		return nil, errors.New("non-canonical bundle base64url payload")
	}
	return decoded, nil
}

func derivePBKDF2SHA256Key(password, salt []byte, iter, keyLen int) []byte {
	numBlocks := (keyLen + sha256.Size - 1) / sha256.Size
	var key []byte
	var blockNum [4]byte

	for i := 1; i <= numBlocks; i++ {
		binary.BigEndian.PutUint32(blockNum[:], uint32(i))

		mac := hmac.New(sha256.New, password)
		mac.Write(salt)
		mac.Write(blockNum[:])
		u := mac.Sum(nil)

		block := make([]byte, len(u))
		copy(block, u)

		for j := 1; j < iter; j++ {
			mac.Reset()
			mac.Write(u)
			u = mac.Sum(nil)
			for k := range block {
				block[k] ^= u[k]
			}
		}

		key = append(key, block...)
	}

	return key[:keyLen]
}

type EncryptResult struct {
	Salt       string
	IV         string
	CT         string
	Iterations int
}

func EncryptWithPassword(plaintext, password string, iterations int) (*EncryptResult, error) {
	if err := validatePassword(password); err != nil {
		return nil, err
	}
	if err := validateIterations(iterations); err != nil {
		return nil, err
	}

	salt := make([]byte, SaltBytes)
	if _, err := rand.Read(salt); err != nil {
		return nil, fmt.Errorf("failed to generate random salt: %w", err)
	}

	iv := make([]byte, IVBytes)
	if _, err := rand.Read(iv); err != nil {
		return nil, fmt.Errorf("failed to generate random iv: %w", err)
	}

	key := derivePBKDF2SHA256Key([]byte(password), salt, iterations, KeyBytes)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("failed to create cipher: %w", err)
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("failed to create gcm: %w", err)
	}

	ct := aesgcm.Seal(nil, iv, []byte(plaintext), nil)

	return &EncryptResult{
		Salt:       toB64(salt),
		IV:         toB64(iv),
		CT:         toB64(ct),
		Iterations: iterations,
	}, nil
}

func DecryptWithPassword(saltB64, ivB64, ctB64 string, iterations int, password string) (string, error) {
	if err := validatePassword(password); err != nil {
		return "", err
	}
	if err := validateIterations(iterations); err != nil {
		return "", err
	}

	salt, err := fromB64(saltB64, "salt")
	if err != nil {
		return "", err
	}
	if len(salt) != SaltBytes {
		return "", fmt.Errorf("invalid salt length: expected %d bytes, got %d", SaltBytes, len(salt))
	}

	iv, err := fromB64(ivB64, "IV")
	if err != nil {
		return "", err
	}
	if len(iv) != IVBytes {
		return "", fmt.Errorf("invalid IV length: expected %d bytes, got %d", IVBytes, len(iv))
	}

	ct, err := fromB64(ctB64, "ciphertext")
	if err != nil {
		return "", err
	}
	if len(ct) < GCMTagBytes {
		return "", errors.New("ciphertext too short (missing GCM tag)")
	}

	key := derivePBKDF2SHA256Key([]byte(password), salt, iterations, KeyBytes)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("failed to create cipher: %w", err)
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("failed to create gcm: %w", err)
	}

	plaintext, err := aesgcm.Open(nil, iv, ct, nil)
	if err != nil {
		return "", fmt.Errorf("decryption failed: %w", err)
	}

	return string(plaintext), nil
}

File: sdks/rust/src/crypto.rs
use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Nonce,
};
use base64::{
    engine::general_purpose::{STANDARD, URL_SAFE_NO_PAD},
    Engine as _,
};
use pbkdf2::pbkdf2_hmac;
use rand::RngCore;
use sha2::Sha256;

use crate::error::{AgentConfigError, Result};
use crate::schema::{ACB_PBKDF2_MAX_ITERATIONS, ACB_PBKDF2_MIN_ITERATIONS};

pub const SALT_BYTES: usize = 16;
pub const IV_BYTES: usize = 12;
pub const KEY_BYTES: usize = 32;
pub const GCM_TAG_BYTES: usize = 16;

pub fn validate_password(password: &str) -> Result<()> {
    if password.is_empty() {
        return Err(AgentConfigError::PasswordError(
            "Password must not be empty".to_string(),
        ));
    }
    Ok(())
}

pub fn validate_iterations(iterations: u32) -> Result<()> {
    if iterations < ACB_PBKDF2_MIN_ITERATIONS || iterations > ACB_PBKDF2_MAX_ITERATIONS {
        return Err(AgentConfigError::ValidationError(format!(
            "PBKDF2 iterations must be between {} and {}",
            ACB_PBKDF2_MIN_ITERATIONS, ACB_PBKDF2_MAX_ITERATIONS
        )));
    }
    Ok(())
}

pub fn to_b64(data: &[u8]) -> String {
    STANDARD.encode(data)
}

pub fn from_b64(value: &str, field_name: &str) -> Result<Vec<u8>> {
    let decoded = STANDARD
        .decode(value)
        .map_err(|e| AgentConfigError::Base64Error(format!("Invalid {field_name}: {e}")))?;
    if to_b64(&decoded) != value {
        return Err(AgentConfigError::Base64Error(format!(
            "Non-canonical {field_name}"
        )));
    }
    Ok(decoded)
}

pub fn to_b64url(data: &[u8]) -> String {
    URL_SAFE_NO_PAD.encode(data)
}

pub fn from_b64url(value: &str) -> Result<Vec<u8>> {
    let decoded = URL_SAFE_NO_PAD
        .decode(value)
        .map_err(|e| AgentConfigError::Base64Error(format!("Invalid base64url: {e}")))?;
    if to_b64url(&decoded) != value {
        return Err(AgentConfigError::Base64Error(
            "Non-canonical bundle base64url payload".to_string(),
        ));
    }
    Ok(decoded)
}

pub fn derive_key(password: &str, salt: &[u8], iterations: u32) -> [u8; KEY_BYTES] {
    let mut key = [0u8; KEY_BYTES];
    pbkdf2_hmac::<Sha256>(password.as_bytes(), salt, iterations, &mut key);
    key
}

pub struct EncryptResult {
    pub salt: String,
    pub iv: String,
    pub ct: String,
    pub iterations: u32,
}

pub fn encrypt_with_password(
    plaintext: &str,
    password: &str,
    iterations: u32,
) -> Result<EncryptResult> {
    validate_password(password)?;
    validate_iterations(iterations)?;

    let mut salt = [0u8; SALT_BYTES];
    rand::thread_rng().fill_bytes(&mut salt);

    let mut iv = [0u8; IV_BYTES];
    rand::thread_rng().fill_bytes(&mut iv);

    let key = derive_key(password, &salt, iterations);
    let cipher = Aes256Gcm::new_from_slice(&key)
        .map_err(|e| AgentConfigError::CryptoError(e.to_string()))?;
    let nonce = Nonce::from_slice(&iv);

    let ciphertext = cipher
        .encrypt(nonce, plaintext.as_bytes())
        .map_err(|e| AgentConfigError::CryptoError(e.to_string()))?;

    Ok(EncryptResult {
        salt: to_b64(&salt),
        iv: to_b64(&iv),
        ct: to_b64(&ciphertext),
        iterations,
    })
}

pub fn decrypt_with_password(
    salt_b64: &str,
    iv_b64: &str,
    ct_b64: &str,
    iterations: u32,
    password: &str,
) -> Result<String> {
    validate_password(password)?;
    validate_iterations(iterations)?;

    let salt = from_b64(salt_b64, "salt")?;
    if salt.len() != SALT_BYTES {
        return Err(AgentConfigError::ValidationError(format!(
            "Invalid salt length: expected {SALT_BYTES} bytes, got {}",
            salt.len()
        )));
    }

    let iv = from_b64(iv_b64, "IV")?;
    if iv.len() != IV_BYTES {
        return Err(AgentConfigError::ValidationError(format!(
            "Invalid IV length: expected {IV_BYTES} bytes, got {}",
            iv.len()
        )));
    }

    let ct = from_b64(ct_b64, "ciphertext")?;
    if ct.len() < GCM_TAG_BYTES {
        return Err(AgentConfigError::ValidationError(
            "Ciphertext too short (missing GCM tag)".to_string(),
        ));
    }

    let key = derive_key(password, &salt, iterations);
    let cipher = Aes256Gcm::new_from_slice(&key)
        .map_err(|e| AgentConfigError::CryptoError(e.to_string()))?;
    let nonce = Nonce::from_slice(&iv);

    let plaintext_bytes = cipher
        .decrypt(nonce, ct.as_ref())
        .map_err(|e| AgentConfigError::CryptoError(format!("Decryption failed: {e}")))?;

    String::from_utf8(plaintext_bytes)
        .map_err(|e| AgentConfigError::ValidationError(format!("Invalid UTF-8 plaintext: {e}")))
}
